File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1: return nums[0]
        def dp(start,end):
            max_sum, adjacent_sum, non_adjacent_sum = 0,0,0
            for i in range(start,end):
                max_sum = max(max_sum,non_adjacent_sum+nums[i])
                non_adjacent_sum = adjacent_sum
                adjacent_sum = max_sum
            return max_sum
        return max(dp(0,len(nums)-1),dp(1,len(nums)))
    
    # a b c d e f g h 
    # 0 1 2 3 4 5 6 7

File: 0213-house-robber-ii/0213-house-robber-ii.cs
public class Solution {
    public int Rob(int[] nums) {
        if(nums.Length == 1) return nums[0];
        return Math.Max(Helper(nums, 0, nums.Length-1), Helper(nums, 1, nums.Length));
    }
    
    public int Helper(int[] nums, int start, int end) {
        List<int> dp = Enumerable.Repeat(0, nums.Length+2).ToList();
        
        for (int i = start; i < end; i++) {
            int curr = nums[i];
            
            // Calculate the maximum amount that can be robbed
            dp[i+2] = Math.Max(dp[i+1], dp[i] + curr);
        }
        
        return dp[end+1];
    }
}
